# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

as well as an

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [1]:
#Imports:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [2]:
#Code:
sat_2017 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/sat_2017.csv')
act_2017 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/act_2017.csv')

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [3]:
#Code:
# This shows the first 10 results for the sat_2017 dataset
sat_2017.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [4]:
# sat_2017 has 51 rows and five columns
sat_2017.shape

(51, 5)

In [5]:
# This gives us the info for sat_2017...such as allowing us to see if there are any null values
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [6]:
# This shows the first 10 results for the act_2017 dataset
act_2017.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

   Composite  
0       21.0  
1       19.2  
2       19.8  
3       19.7  
4       19.4  
5       22.8  
6       20.8  
7       25.2  
8       24.1  
9       24.2

In [7]:
# act_2017 has 52 rows and 7 columns
act_2017.shape

(52, 7)

In [8]:
# # This gives us the info for act_2017...such as allowing us to see if there are any null values
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   English        52 non-null     float64
 3   Math           52 non-null     float64
 4   Reading        52 non-null     float64
 5   Science        52 non-null     float64
 6   Composite      52 non-null     float64
dtypes: float64(5), object(2)
memory usage: 3.0+ KB


In [9]:
act_2017[1:].describe()

English       Math    Reading    Science  Composite
count  51.000000  51.000000  51.000000  51.000000  51.000000
mean   20.931373  21.182353  22.013725  21.041176  21.519608
std     2.353677   1.981989   2.067271   3.182463   2.020695
min    16.300000  18.000000  18.100000   2.300000  17.800000
25%    19.000000  19.400000  20.450000  19.900000  19.800000
50%    20.700000  20.900000  21.800000  21.300000  21.400000
75%    23.300000  23.100000  24.150000  22.750000  23.600000
max    25.500000  25.300000  26.000000  24.900000  25.500000

In [10]:
act_2017

State Participation  English  Math  Reading  Science  \
0               National           60%     20.3  20.7     21.4     21.0   
1                Alabama          100%     18.9  18.4     19.7     19.4   
2                 Alaska           65%     18.7  19.8     20.4     19.9   
3                Arizona           62%     18.6  19.8     20.1     19.8   
4               Arkansas          100%     18.9  19.0     19.7     19.5   
5             California           31%     22.5  22.7     23.1     22.2   
6               Colorado          100%     20.1  20.3     21.2     20.9   
7            Connecticut           31%     25.5  24.6     25.6     24.6   
8               Delaware           18%     24.1  23.4     24.8     23.6   
9   District of Columbia           32%     24.4  23.5     24.9     23.5   
10               Florida           73%     19.0  19.4     21.0     19.4   
11               Georgia           55%     21.0  20.9     22.0     21.3   
12                Hawaii           90%     17.8  19.2     19.2     19.3   
13                 Idaho           38%     21.9  21.8     23.0     22.1   
14              Illinois           93%     21.0  21.2     21.6     21.3   
15               Indiana           35%     22.0  22.4     23.2     22.3   
16                  Iowa           67%     21.2  21.3     22.6     22.1   
17                Kansas           73%     21.1  21.3     22.3     21.7   
18              Kentucky          100%     19.6  19.4     20.5     20.1   
19             Louisiana          100%     19.4  18.8     19.8     19.6   
20                 Maine            8%     24.2  24.0     24.8     23.7   
21              Maryland           28%     23.3  23.1     24.2      2.3   
22         Massachusetts           29%     25.4  25.3     25.9     24.7   
23              Michigan           29%     24.1  23.7     24.5     23.8   
24             Minnesota          100%     20.4  21.5     21.8     21.6   
25           Mississippi          100%     18.2  18.1     18.8     18.8   
26              Missouri          100%     19.8  19.9     20.8     20.5   
27               Montana          100%     19.0  20.2     21.0     20.5   
28              Nebraska           84%     20.9  20.9     21.9     21.5   
29                Nevada          100%     16.3  18.0     18.1     18.2   
30         New Hampshire           18%     25.4  25.1     26.0     24.9   
31            New Jersey           34%     23.8  23.8     24.1     23.2   
32            New Mexico           66%     18.6  19.4     20.4     20.0   
33              New York           31%     23.8  24.0     24.6     23.9   
34        North Carolina          100%     17.8  19.3     19.6     19.3   
35          North Dakota           98%     19.0  20.4     20.5     20.6   
36                  Ohio           75%     21.2  21.6     22.5     22.0   
37              Oklahoma          100%     18.5  18.8     20.1     19.6   
38                Oregon           40%     21.2  21.5     22.4     21.7   
39          Pennsylvania           23%     23.4  23.4     24.2     23.3   
40          Rhode Island           21%     24.0  23.3     24.7     23.4   
41        South Carolina          100%     17.5  18.6     19.1     18.9   
42          South Dakota           80%     20.7  21.5     22.3     22.0   
43             Tennessee          100%     19.5  19.2     20.1     19.9   
44                 Texas           45%     19.5  20.7     21.1     20.9   
45                  Utah          100%     19.5  19.9     20.8     20.6   
46               Vermont           29%     23.3  23.1     24.4     23.2   
47              Virginia           29%     23.5  23.3     24.6     23.5   
48            Washington           29%     20.9  21.9     22.1     22.0   
49         West Virginia           69%     20.0  19.4     21.2     20.5   
50             Wisconsin          100%     19.7  20.4     20.6     20.9   
51               Wyoming          100%     19.4  19.8     20.8     20.6   

    Composite  
0        21.0  
1        19

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

**Answer:**
First we will start with the sat_2017 dataset. This dataset contains 51 rows for all 50 states and the District of Columbia. There are 5 columns in this data set: 'State', 'Participation', Evidence-Based Reading and Writing', 'Math', and 'Total.' In sat_2017 there are no null values which means the dataset is complete. The average scores across all 50 states and the District of Columbia for Evidence-Based Reading and Writing is a score of 569. The average scores across all 50 states and the District of Columbia for Math is a score of 547. The average scores across all 50 states and the District of Columbia for the toal is a score of 1126.

The act_2017 dataset contains 52 rows. This includes 50 rows for all 50 states, a row for the District of Columbia and a row for the nation as a whole. There are 7 columns in this dataset: 'State', 'Participation', 'English', 'Math', 'Reading', 'Science', 'Composite.'In act_2017 there are no null values which means the dataset is complete. However in the 'Composite' column Wyoming has a value of 20.2x which is not like the other values so this was most likely an error which occured in data entry. In the 'Science' column Maryland has a value of 2.3 which is not possible. The average score for English was 20.9, the average socre for Math was 21.1, the average score for reading ws 22.0, the average score for science was 21.0. 

#### 4a. Does the data look complete? 

**Answer:** No, the data is not complete. There are a few errors in both datasets. 


The sat_2017 dataset is complete. The act_2017 dataset is not complete because in the 'Composite' column Wyoming has a value of 20.2x making this dataset not complete. In the act_2017 dataset 

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

**Answer:**

sat_2017
* **ERROR** Accoring to https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/ the value for Maryland in the 'Math' column should be 524.
* The minimum for Evidence-Base Reading and riding was 482. The minimum for the total was 950. The maximum for Math is 651. 
* The maximum for Evidence-Based Reading and Writing is 644. The maximum for total is 1295.

act_2017

* **ERROR** Accoring to https://www.act.org/content/dam/act/unsecured/documents/cccr2017/ACT_2017-Average_Scores_by_State.pdf the value for Maryland in the 'Science' column should be 23.3 not 2.3
* **ERROR** In the 'Composite' column Wyoming has a value of 20.2x. It should just be 20.2
* The minimum value in the 'Science' column is 2.3. The minimum for 'English' is 16.3. The minimum for 'Math' is 18.0. The minimum for 'Reading' is 18.1. 
* The maximum for 'English' is 25.5. The maximum for 'Math' is 25.3. The maximum for 'Reading' is 26. The maximum for 'Science' is 24.9.

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

**ANSWER:** 


In [165]:
# How to fix the errors I identified

# This fixes the error of Maryland in the column 'math_sat' Changes 52 to 524
sat_2017.at[20, 'math_sat'] = 524

# This fixes the error for Maryland in the column 'science_act' Changes 2.3 to 23.2
act_2017.at[21, 'science_act']= 23.2

# This fixes the error for Wyoming in the column 'composite' Changes 20.2x to 20.2
act_2017.at[51, 'composite_act']= 20.2

#### 5. What are your data types? 
Display the data types of each feature. 

In [166]:
sat_2017.info()
# 'State' is an object, 'Participation' is an object, 'Evidence-Based Reading and Writing' is an integer,
# 'Math' is an integer, 'Total' is an integer

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   state_sat                               51 non-null     object 
 1   participation_sat                       51 non-null     float64
 2   evidence-based reading and writing_sat  51 non-null     int64  
 3   math_sat                                51 non-null     int64  
 4   total_sat                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [167]:
act_2017.info()
# 'State' is an object, 'Participation' is an object, 'English' is an float, 'Math' is an float, 
# 'Science' is an float, 'Composite' was an object until we updated the value now it is an float

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   state_act          52 non-null     object 
 1   participation_act  52 non-null     float64
 2   english_act        52 non-null     float64
 3   math_act           52 non-null     float64
 4   reading_act        52 non-null     float64
 5   science_act        52 non-null     float64
 6   composite_act      52 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

**Answer:** 
* 'Composite' in act_2017 was not correct it should have been a float like the others. We updated that value, now it is correct. 
*  Participation in both datasets should also be floats. This would allow us to use those values further in research. We updated both 'Participation' columns, now they are correct.

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [150]:
# https://stackoverflow.com/questions/15891038/change-data-type-of-columns-in-pandas
# https://towardsdatascience.com/understand-map-function-to-manipulate-pandas-series-8ac340d514f7

In [149]:
# This shows that this is a series
# This is the value that we need to change. It wa originally an object and it should be a float.
type(sat_2017['participation_sat'])

pandas.core.series.Series

In [79]:
# This is an object that needs to be a float
sat_2017['Participation'].tail()

46    65%
47    64%
48    14%
49     3%
50     3%
Name: Participation, dtype: object

In [296]:
# 'Participation' has object values like '5%' we need to change this to a float.
# We need to drop the % from every string


# It is throwing an error since I ran all kernels it is trying to map that function to the 'Participation' column
# where that function is already applied

# sat_2017['Participation'] = sat_2017['Participation'].map(lambda x: float(x[:-1]))

In [124]:
# We can see here due to that function participation is now a float
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   state_sat                               51 non-null     object 
 1   participation_sat                       51 non-null     float64
 2   evidence-based reading and writing_sat  51 non-null     int64  
 3   math_sat                                51 non-null     int64  
 4   total_sat                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [125]:
# We can see here due to that function participation is now a float
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   state_act          52 non-null     object 
 1   participation_act  52 non-null     float64
 2   english_act        52 non-null     float64
 3   math_act           52 non-null     float64
 4   reading_act        52 non-null     float64
 5   science_act        52 non-null     float64
 6   composite_act      52 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


In [126]:
# The values in participation do not have the '%' now
act_2017.tail()

state_act  participation_act  english_act  math_act  reading_act  \
47       Virginia               29.0         23.5      23.3         24.6   
48     Washington               29.0         20.9      21.9         22.1   
49  West Virginia               69.0         20.0      19.4         21.2   
50      Wisconsin              100.0         19.7      20.4         20.6   
51        Wyoming              100.0         19.4      19.8         20.8   

    science_act  composite_act  
47         23.5           23.8  
48         22.0           21.9  
49         20.5           20.4  
50         20.9           20.5  
51         20.6           20.2

In [168]:
# This changes the values in 'Participation' to a float

# It is throwing an error since I ran all kernels it is trying to map that function to the 'Participation' column
# where that function is already applied

# This is the function that changed the values in the 'Participation' column to floats
# act_2017['Participation'] = act_2017['Participation'].map(lambda x: float(x[:-1]))

In [128]:
# The values in 'Participation' are now floats
# The values in participation do not have the '%' now
act_2017.tail()

state_act  participation_act  english_act  math_act  reading_act  \
47       Virginia               29.0         23.5      23.3         24.6   
48     Washington               29.0         20.9      21.9         22.1   
49  West Virginia               69.0         20.0      19.4         21.2   
50      Wisconsin              100.0         19.7      20.4         20.6   
51        Wyoming              100.0         19.4      19.8         20.8   

    science_act  composite_act  
47         23.5           23.8  
48         22.0           21.9  
49         20.5           20.4  
50         20.9           20.5  
51         20.6           20.2

In [136]:
act_2017.iloc[51]

state_act            Wyoming
participation_act        100
english_act             19.4
math_act                19.8
reading_act             20.8
science_act             20.6
composite_act           20.2
Name: 51, dtype: object

- Fix any individual values preventing other columns from being the appropriate type.

In [169]:
#code
# Since I was able to change the value of Wyoming in the 'Composite' column in the act_2017.csv this updated the
# Datatype of the coulmn since I removed the x from 20.2x making the value an integer.


# This is the code that would be used to update the Wyoming value in the 'Composite' column
# act_2017.at[51, 'composite_act']= 20.2

- Finish your data modifications by making sure the columns are now typed appropriately.

In [170]:
#code
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   state_act          52 non-null     object 
 1   participation_act  52 non-null     float64
 2   english_act        52 non-null     float64
 3   math_act           52 non-null     float64
 4   reading_act        52 non-null     float64
 5   science_act        52 non-null     float64
 6   composite_act      52 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


In [171]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   state_sat                               51 non-null     object 
 1   participation_sat                       51 non-null     float64
 2   evidence-based reading and writing_sat  51 non-null     int64  
 3   math_sat                                51 non-null     int64  
 4   total_sat                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


- Display the data types again to confirm they are correct.

In [172]:
#Code:
# All columns in act_2017 are the datatypes they should be
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   state_act          52 non-null     object 
 1   participation_act  52 non-null     float64
 2   english_act        52 non-null     float64
 3   math_act           52 non-null     float64
 4   reading_act        52 non-null     float64
 5   science_act        52 non-null     float64
 6   composite_act      52 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


In [173]:
# All the columns in sat_2017 are the datatypes they should be
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   state_sat                               51 non-null     object 
 1   participation_sat                       51 non-null     float64
 2   evidence-based reading and writing_sat  51 non-null     int64  
 3   math_sat                                51 non-null     int64  
 4   total_sat                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [176]:
#code
# This updates the columns for each dataset

# We add _act to each column name to make it more distinguishable against sat_2017
act_2017.columns = ['state_act', 'participation_act', 'english_act', 'math_act', 'reading_act', 'science_act', 'composite_act']

# We add _sat to each column name to make it more distinguishable against sact_2017
sat_2017.columns = ['state_sat', 'participation_sat', 'evidence_based_reading_and_writing_sat', 'math_sat', 'total_sat']

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**state_sat**|object|sat_2017|All 50 states and the District of Columbia|
|**participation_sat**|float|sat_2017|The overall percentage of people taking the SAT in that state in 2017|
|**evidence_based_reading_and_writing_sat**|int|sat_2017|The average score for the evidence-based reading and writing section of the SAT for each state |
|**math_sat**|int|sat_2017|The average score for the math section of the SAT for each state|
|**total_sat**|int|sat_2017|The overall average total score of the SAT for each state |
|**state_act**|object|act_2017|All 50 states, the District of Columbia, and the national average for each section|
|**participation_act**|float|act_2017|The overall percentage of people taking the ACT in that state in 2017|
|**english_act**|float|act_2017|The average score for the english section of the ACT for each state and the national average|
|**math_act**|float|act_2017|The average score for the math section of the ACT for each state and the national average|
|**reading_act**|float|act_2017|The average score for the reading section of the ACT for each state and the national average |
|**science_act**|float|act_2017|The average score for the science section of the ACT for each state and the national average |
|**composite_act**|float|act_2017|The average score on the ACT for each state and the national average score|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [214]:
# This is how I dropped the National row
# Keep in comment or it will continute to delete rows
# act_2017.drop(act_2017.index[[0]])

In [215]:
# This is how to add a new row
# data = []
# data.insert(0, {'state_act': 'Alabama', 'participation_act': 100.0, 'english_act': 18.9, 'math_act': 18.4, 'reading_act': 19.7, 'science_act': 19.4, 'composite_act': 19.2}
# act_2017 = pd.concat([pd.DataFrame(data), act_2017], ignore_index=True)

In [216]:
# 51 rows and 7 columns for the act_2017 dataset
act_2017.shape

(51, 7)

In [218]:
# 51 rows and 5 columns for the sat_2017 dataset
sat_2017.shape

(51, 5)

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [23]:
#Code:

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [24]:
#code

## 2018 Data Import and Cleaning

Read in the `sat_2018.csv` and `act_2018.csv` files and assign them to appropriately named pandas dataframes. For the **2018 ACT Data**, only the `Composite` scores are available. Repeat the same processes to clean the 2018 data here as you were instructed in the previous sections above.

In [259]:
sat_2018 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/sat_2018.csv')
act_2018 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/act_2018.csv')

# Had to use this since I accidently assigned the sat_2018['Participation'] = 'act_2018'['Particpation'] when changing 
# The values from objects to float
sat_20182 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/sat_2018.csv')

In [229]:
# The amount of rows and columns are correct here
sat_2018.shape

(51, 5)

In [230]:
# act_2018 has an extra row
# There are two rows with state Maine: 19 and 20
act_2018.shape

(52, 3)

In [240]:
# There are no values that seem out of place or not possible
sat_2018.describe()

Evidence-Based Reading and Writing        Math        Total
count                           51.000000   51.000000    51.000000
mean                           563.686275  556.235294  1120.019608
std                             47.502627   47.772623    94.155083
min                            480.000000  480.000000   977.000000
25%                            534.500000  522.500000  1057.500000
50%                            552.000000  544.000000  1098.000000
75%                            610.500000  593.500000  1204.000000
max                            643.000000  655.000000  1298.000000

In [243]:
 # There are no values that seem out of place or not possible
act_2018.describe()

Composite
count  52.000000
mean   21.544231
std     2.119417
min    17.700000
25%    19.975000
50%    21.300000
75%    23.725000
max    25.600000

In [247]:
# Participation has to become a float
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [248]:
# Participation has to become a float
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   Composite      52 non-null     float64
dtypes: float64(1), object(2)
memory usage: 1.3+ KB


In [260]:
# This changes the values in 'Participation' to floats
sat_2018['Participation'] = sat_20182['Participation'].map(lambda x: float(x[:-1]))

In [262]:
# 'Participation' datatype is now a float
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   State                               51 non-null     object 
 1   Participation                       51 non-null     float64
 2   Evidence-Based Reading and Writing  51 non-null     int64  
 3   Math                                51 non-null     int64  
 4   Total                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [264]:
# This changes the values in 'Participation' to floats
act_2018['Participation'] = act_2018['Participation'].map(lambda x: float(x[:-1]))

In [265]:
# 'Participation' datatype is now a float
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     float64
 2   Composite      52 non-null     float64
dtypes: float64(2), object(1)
memory usage: 1.3+ KB


In [268]:
# Now I need to get rid of the error
# There is an extra row in the act_2018 dataset. Maine was put in twice
act_2018.shape

(52, 3)

In [273]:
# This is how we get rid of duplicate rows
act_2018 = act_2018.drop_duplicates()

In [279]:
# The act_2018 dataframe now has 51 rows instead of 52 because we took out the duplicate Maine
act_2018.shape

(51, 3)

In [285]:
# Now we have to change the column name...this will make our lives easier down the line

# This is how to update the act_2018 columns
act_2018.columns = ['state_act', 'participation_act', 'composite_act']

# This is how to update the sat_2018 columns
sat_2018.columns = ['state_sat', 'participation_sat', 'evidence_based_reading_and_writing_sat', 'math_sat', 'total_sat']

In [295]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   state_sat                               51 non-null     object 
 1   participation_sat                       51 non-null     float64
 2   evidence_based_reading_and_writing_sat  51 non-null     int64  
 3   math_sat                                51 non-null     int64  
 4   total_sat                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [297]:
sat_2018

state_sat  participation_sat  \
0                Alabama                6.0   
1                 Alaska               43.0   
2                Arizona               29.0   
3               Arkansas                5.0   
4             California               60.0   
5               Colorado              100.0   
6            Connecticut              100.0   
7               Delaware              100.0   
8   District of Columbia               92.0   
9                Florida               56.0   
10               Georgia               70.0   
11                Hawaii               56.0   
12                 Idaho              100.0   
13              Illinois               99.0   
14               Indiana               63.0   
15                  Iowa                3.0   
16                Kansas                4.0   
17              Kentucky                4.0   
18             Louisiana                4.0   
19                 Maine               99.0   
20              Maryland               76.0   
21         Massachusetts               80.0   
22              Michigan              100.0   
23             Minnesota                4.0   
24           Mississippi                3.0   
25              Missouri                4.0   
26               Montana               10.0   
27              Nebraska                3.0   
28                Nevada               23.0   
29         New Hampshire               96.0   
30            New Jersey               82.0   
31            New Mexico               16.0   
32              New York               79.0   
33        North Carolina               52.0   
34          North Dakota                2.0   
35              Oklahoma                8.0   
36                  Ohio               18.0   
37                Oregon               48.0   
38          Pennsylvania               70.0   
39          Rhode Island               97.0   
40        South Carolina               55.0   
41          South Dakota                3.0   
42             Tennessee                6.0   
43                 Texas               66.0   
44                  Utah                4.0   
45               Vermont               64.0   
46              Virginia               68.0   
47            Washington               69.0   
48         West Virginia               28.0   
49             Wisconsin                3.0   
50               Wyoming                3.0   

    evidence_based_reading_and_writing_sat  math_sat  total_sat  
0                                      595       571       1166  
1                                      562       544       1106  
2                                      577       572       1149  
3                                      592       576       1169  
4                                      540       536       1076  
5                                      519       506       1025  
6                                      535       519       1053  
7                                      505       492        998  
8                                      497       480        977  
9                                      550       549       1099  
10                                     542       522       1064  
11                                     480       530       1010  
12                                     508       493       1001  
13                                     513       506       1019  
14                                     542       532       1074  
15                                     634       631       1265  
16                                     633       631       1265  
17                                     630       618       1248  
18                                     615       595       1210  
19                                     512       501       1013  
20                                     545       535       1080  
21                                     562       563       1125  
22                                     511       499

# Data dictionary for ACT and SAT data in 2018
|Feature|Type|Dataset|Description|
|---|---|---|---|
|**state_sat**|object|sat_2018|All 50 states and the District of Columbia|
|**participation_sat**|float|sat_2018|The overall percentage of people taking the SAT in that state in 2018|
|**evidence_based_reading_and_writing_sat**|int|sat_2018|The average score for the evidence-based reading and writing section of the SAT for each state |
|**math_sat**|int|sat_2018|The average score for the math section of the SAT for each state|
|**total_sat**|int|sat_2018|The overall average total score of the SAT for each state |
|**state_act**|object|act_2018|All 50 states and the District of Columbia|
|**participation_act**|float|act_2018|The overall percentage of people taking the ACT in that state in 2018|
|**composite_act**|float|act_2018|The average score on the ACT for each state and the national average score|

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [25]:
#Code:

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [26]:
#code

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [27]:
#Code:

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

Answer

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [28]:
#code

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [29]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [30]:
# Code

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [31]:
# Code

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [32]:
# Code

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

## Bonus: Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer:

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [33]:
# Code: